File: packages/app/modules/paseo-word-stream/ios/WordStreamModule.swift
import ExpoModulesCore
import UIKit

struct WordRangeRecord: Record {
  @Field var start: Int = 0
  @Field var end: Int = 0
  @Field var startedAt: Double = 0
  @Field var spanMs: Double = 0
}

public final class WordStreamModule: Module {
  public func definition() -> ModuleDefinition {
    Name("PaseoWordStream")
    View(WordFadeHost.self) {
      Prop("ranges") { (view: WordFadeHost, ranges: [WordRangeRecord]) in
        view.ranges = ranges.map {
          WordFadeRange(start: $0.start, end: $0.end, startedAt: $0.startedAt, spanMs: $0.spanMs)
        }
      }
    }
  }
}

final class WordFadeHost: ExpoView {
  private weak var textView: UITextView?
  private var animator: TailFadeInAnimator?
  var ranges: [WordFadeRange] = [] {
    didSet {
      animator?.receive(ranges: ranges)
      setNeedsLayout()
    }
  }

  override func layoutSubviews() {
    super.layoutSubviews()
    // RN's child owns geometry. Its drawRect installs the attributed string.
    subviews.forEach { $0.layer.displayIfNeeded() }
    bindTextView()
  }

  override func didAddSubview(_ subview: UIView) {
    super.didAddSubview(subview)
    setNeedsLayout()
  }

  override func willRemoveSubview(_ subview: UIView) {
    animator?.close()
    animator = nil
    textView = nil
    super.willRemoveSubview(subview)
  }

  override func didMoveToWindow() {
    super.didMoveToWindow()
    if window == nil {
      animator?.close()
      animator = nil
      textView = nil
    } else {
      setNeedsLayout()
    }
  }

  override func prepareForRecycle() {
    ranges = []
    animator?.close()
    animator = nil
    textView = nil
    super.prepareForRecycle()
  }

  private func bindTextView() {
    guard let child = Self.findTextView(self) else { return }
    if child !== textView {
      animator?.close()
      textView = child
      animator = TailFadeInAnimator(textView: child)
    }
    animator?.receive(ranges: ranges)
  }

  private static func findTextView(_ view: UIView) -> UITextView? {
    if let text = view as? UITextView { return text }
    for child in view.subviews {
      if let text = findTextView(child) { return text }
    }
    return nil
  }
}

File: packages/app/modules/paseo-word-stream/android/src/main/java/sh/paseo/wordstream/WordStreamModule.kt
package sh.paseo.wordstream

import android.content.Context
import android.view.View
import android.view.ViewTreeObserver
import android.widget.TextView
import android.view.ViewGroup
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Record
import expo.modules.kotlin.records.Field

class WordRange : Record {
  @Field var start: Int = 0
  @Field var end: Int = 0
  @Field var startedAt: Double = 0.0
  @Field var spanMs: Double = 0.0
}

class WordStreamModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("PaseoWordStream")
    View(WordFadeHost::class) {
      Prop("ranges") { view: WordFadeHost, ranges: List<WordRange> -> view.ranges = ranges }
      OnViewDestroys { view: WordFadeHost -> view.reset() }
    }
  }
}

/** Fabric owns child layout and commits the text and ranges before pre-draw. */
class WordFadeHost @JvmOverloads constructor(
  context: Context,
  private val frameClock: WordFadeFrameClock = AndroidWordFadeFrameClock,
) : ViewGroup(context) {
  override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {}

  private var animator: TailFadeInAnimator? = null
  var ranges: List<WordRange> = emptyList()
    set(value) {
      field = value
      animator?.cancelAll()
      applyRanges()
      invalidate()
    }
  private val beforeDraw = ViewTreeObserver.OnPreDrawListener { applyRanges(); true }
  private var observer = viewTreeObserver

  init { observer.addOnPreDrawListener(beforeDraw) }

  override fun onViewAdded(child: View) {
    super.onViewAdded(child)
    require(child is TextView) { "WordFadeHost requires a single root TextView" }
    require(childCount == 1) { "WordFadeHost requires a single child" }
    animator = TailFadeInAnimator(child, frameClock)
    applyRanges()
  }

  override fun onViewRemoved(child: View) {
    animator?.cancelAll()
    animator = null
    super.onViewRemoved(child)
  }

  private fun applyRanges() { animator?.receive(ranges) }

  override fun onAttachedToWindow() {
    super.onAttachedToWindow()
    observer = viewTreeObserver
    observer.removeOnPreDrawListener(beforeDraw)
    observer.addOnPreDrawListener(beforeDraw)
    applyRanges()
  }

  override fun onDetachedFromWindow() {
    if (observer.isAlive) observer.removeOnPreDrawListener(beforeDraw)
    animator?.cancelAll()
    super.onDetachedFromWindow()
  }

  fun reset() {
    ranges = emptyList()
    animator?.cancelAll()
  }
}
